### DISCUSSION TITLE
Shared DB

### DISCUSSION THREAD

**Discussion by thielj** [lldap/lldap#901] [Q&A] [answered]
Hi

Can multiple instances of LLDAP share the same SQLite database through a shared volume or network drive?

How is the situation when using a shared Postgres DB?

It would be sufficient if one LLDAP is used as a "master" to make modifications and the others are used as read-only.

**Comment by jakob42** [comment]
I'm using sqlite, inotify and rsync. It's a cursed setup that works so far. 🫣

**Reply by jakob42** [reply]
But I'm fully aware that there is no support for my setup

**Reply by nitnelave** [reply]
That's a bit of a risky setup, but as long as you don't have concurrent writes, it should be fine.

**Reply by jakob42** [reply]
Sorry, forgot to mention, I only sync from one server to the two others. 

And yeah, I bet it's risky, but it worked for about 6 month now.

**Reply by nitnelave** [reply]
If you only sync one way, it should be safe (though you of course lose the writes on the other 2 servers). You could look into SQLite replication solutions, some are mentioned here: https://sqlite.org/forum/info/4fc617ecaf47d544ea8baae0930553a3a52e63067d3adfff9ce1a217cc470aba

**Reply by thielj** [reply]
Hi Jakob. You may want to have a look at one of these:

https://github.com/benbjohnson/litestream -
https://github.com/rqlite/rqlite

**Reply by peter-** [reply]
FWIW, I'm ended up using [litefs](https://fly.io/docs/litefs/) since litestream "[decided to \[remain\] a pure disaster recovery tool and not include live replication](https://github.com/benbjohnson/litestream/issues/8#issuecomment-1173214316)".

In my setup only one server has a writable sqlite db (which is also the only server I chose to expose the lldap web UI on), all other lldap instances open their "clustered" sqlite dbs in read-only mode:
```toml
database_url = "sqlite:///litefs/fuse/users.db?mode=ro"
```

With the minor nuisance of error messages in the logs from clean up jobs that wouldn't have to be run in the first place (`DB error while cleaning up` due to an `attempt to write a readonly database`) those "replica" lldap instances seem to work just fine.

(I suppose telling lldap the db is supposed to be read-only would help avoid those log messages and could disable any jobs that need write access to the db at the same time. Filed as #1236.)

**Comment by nitnelave** [comment, marked as answer]
Each LLDAP instance is stateless. Assuming they all have the same config, sharing a postgres or MySQL DB should work well (and afaik many users do this, as a HA setup).

Make sure to use a key_seed, not a key_file for the private key.

Sharing SQLite is not recommended (each file should only be opened by a single process).

**Reply by thielj** [reply]
HI. I think SQLite is safe to be used by multiple processes.

https://www.sqlite.org/faq.html#q5

**Reply by peter-** [reply]
Maybe also let #1236 serve as an indicator that this is all very much possible and can be safe depending on deployment details.
